### ISSUE TITLE
Proposal: Partitioning feature for Multi-Tenancy supporting millions of tenants

### ISSUE THREAD

**Issue by etiennedi** [weaviate/weaviate#2970]
# Proposal: Partitioning for Multi-Tenancy

A multi-tenancy solution that keeps each tenant’s data separate and allows for scaling to millions of tenants. Each tenant uses a dedicated partition.

## Highlights:

* supports millions of partitions overall, with up to 100,000 partitions per node
* easy to use, new partitions will be created automatically 
* full isolation between partitions, easy offboarding of entire partitions
* partitions can be vastly different: Some very small and others very large without performance penalties or other drawbacks
* Weaviate automatically manages the distribution of partitions in the cluster

## Key design features:

* **Non-breaking to existing users**
    * Partitioning is an optional feature. If not used, Weaviate behaves exactly as before.
* **Strong isolation**
    * When using partitioning, Weaviate creates one partition per tenant. A partition (internally a shard) is an isolation unit that is scheduled on exactly one node.
* **Self-contained partitions for efficient queries**
    * Each shard is fully self-contained. No filters are required to isolate queries to a shard. This allows tenants of different sizes to co-exist on the same nodes without performance penalties.
* **Various Compliance Features**
    * Weaviate prevents access across tenants; any action is always scoped to exactly one partition. 
    * Offboarding a tenant is simple and computationally trivial. It simply drops files. No indexes need to be rebuilt or repaired
* **Horizontal Scaling**
    * Add more compute and storage for more partitions, simply by adding more nodes to your cluster
    * Optionally redistribute existing partitions across the cluster
* **Inactive tenants**
    * Only use expensive compute resources for tenants that are currently active.
    * Inactive tenants require fewer resources
    * Optionally offload tenants that have been inactive for a long time to cheap cloud storage
* **Easy to use**
    * The tenant is specified using a partition value as a string
    * If a partition does not exist yet, Weaviate will automatically create it
    * All CRUD-like operations require specifying the partitioning key

### API Summary

* When partitioning is enabled, the existing `shardingConfig` is no longer in use. Instead of a physical&lt;>virtual shard mapping with a fixed number of shards, Weaviate dynamically provisions shards to match the desired partitions (one partition == one shard)
* When partitioning is enabled, any object in a partitioned class must always be specified with the partition value. 
* This includes referencing an object through a cross-reference of a partitioned class

### API Details

Note that this is the very first draft. Your feedback and input is welcome. If you believe this can be improved, please leave a comment below. 

#### Schema

The motivation is to tell Weaviate that a specific class should use partitioning.

```js
{
  /* ... other fields ... */
  "multiTenancyConfig": {
    "tenantKey": "name of the class property to be used as tenant key",
    "enabled": true|false
  },
  "shardingConfig": null,
}

```

In the initial implementation, the `tenantKey` will be restricted to properties of type `text` and `uuid`. This can be expanded to include other property types in future iterations.

#### CRUD Operations (REST)

When CRUDing an object, Weaiviate needs to know which partition the object belongs to. As an example an operation to create a new object:

```sh
curl \
    -X POST \
    -H "Content-Type: application/json" \
    -d '{
      "class": "Author",
      "tenant": "tenant_123" <----------------------------
      "properties": {
          "name": "Jodi Kantor",
       },
      "id": "36ddd591-2dee-4e7e-a3cc-eb86d30a4303"
    }' \
    http://localhost:8080/v1/objects?consistency_level=ALL
```

#### Search Operations (GraphQL/gRPC)

Any operation (listing object, searching objects, etc) need to specify the partition value:

```graphql
{
  Get {
    Author(
      tenant: "tenant_123",
      limit: 42,
      nearText: { /* ... */ } 
    ){
       name
    }
  }
}
```

#### Cross-References

```js
{
  "beacon": "weaviate://localhost/{ClassName}/{uuid}?tenant=tenant_123
}
```

Note: A possible improvement here could be that if both source and destination classes use partitions, that the partition key can be omitted and then the same partition is used on both.

### Vast number of partitions (per node)

The biggest technical hurdle in this proposal is the number of partitions we want to support. In order to support millions of partitions in a reasonably sized cluster, we need to support a large number of partitions per node. The current design goal is 100,000 per node. This is rather arbitrary, but it allows for reasonable cluster sizes, such as 10 nodes for 1 million tenants or 30 for 3 million tenants. 

A partition under the hood is essentially a Weaviate shard. We have run various experiments to identify what a realistic number would be. At the moment, the only way to create many shards is to create many classes. This approach does not scale well (e.g., because of GraphQL). Nevertheless, in a test scenario, we could reach over 30,000 shards on a single node. Note that this is _before_ we optimized to make shards more lightweight. 

The key optimizations that we need to tackle are:

* Reduce the memory footprint of a shard. The current proposal predicts around 500kB of memory per empty shard. This value can be reduced further if necessary.
* Reduce the amount of async processes per shard. Right now, there are about 20 async processes per shard – even with just one user-defined property. This led to around 800,000 goroutines and the same amount of timers in an experiment. This can be reduced to just one per shard
* Unify stores in a shard: A shard with many properties currently leads to many open files and active memory mappings. We can reduce this number, so we can comfortably handle 100,000 shards per node with the commonly found hard file limits on GKE or EKS. 
* Note that the 100,000 shards per node number assumes relatively small partitions. If partitions grow considerably larger the amount of active partitions per node will be lower. We believe this is not a problem in reality, as large partitions will require considerable resources anyway. So, we predict that storage, memory, and compute will be the limiting factor in this cases. With a typical Power Law distribution where there are a few large tenants and a lot of small tenants, it could be entirely feasible to have a single node with just 100 partitions, but another node with 10s of thousands. It is vital that Weaviate can distribute partitions intelligently and rebalance them if needed.

### CP-Nature of Sharding State

Currently, the sharding state is stored as part of the schema. This is CP (in CAP terms) and currently requires all nodes to be healthy to update it. This is not feasible in the long-term with very large clusters. We need to assume that nodes could be down at any time. The effects of serving traffic for existing tenants can be caught by replication. However, with the current design it would be impossible to add or remove a tenant in this case. Given the frequency of onboarding (for example, because an app goes viral), this is not good enough. \
 \
The sharding state will be much simpler. The “partition value to shard name” can be set deterministically, for example, with a hashing function. There is no more needs or virtual shards, as we can rule out that a shard will ever have to be split. Instead, the use case can be scaled by adding more partitions (and more nodes if existing nodes can’t fit the new partitions). Nevertheless, the shard&lt;>node association is extremely important state to the cluster, as it is used to tell each node where to redirect traffic.

Possible remedies to efficiently distribute this state in the cluster could be a distributed caching system similar to DNS or electing specific nodes as leaders for this state. Most likely, this problem will be overcome in the second iteration of this feature (see Timeline below).

## Timeline

As with other large new Weaviate features, this feature will be rolled out over multiple releases. The main functionality will be contained in the first release with additional improvements later on. The current estimated timeline is

* Step 1: Core functionality in v1.20
    * All APIs available
    * supports 100s of thousands of partitions
    * add more resources by adding new nodes to the cluster
* Step 2: More dynamic provisioning in v1.21
    * Move idle partitions to inactive state to save resources
    * Move existing partitions between nodes, rebalance, etc.
    * Improved distribution of shard&lt;>node state
* Step 3: Separation of storage and compute in v1.22
    * Remove inactive partitions from expensive “hot” nodes and offload them to cloud storage
    * Pull in partitions on demand from cloud storage

## Request for Feeback

1. Looking at your use-case, does this satisfy the requirements for multi-tenancies wrt isolation, compliance, and scalability?
2. API feedback including the naming of parameters
3. Is anything missing or would you like to change something?

**Comment by byronvoorbach** [comment]
I think it’s a very good writeup and a nice proposal. I agree with everything written, a few topics come to mind:

- How will we handle backups in these kinds of cases? 
    - The current backup functionality will not work, as it would backup the entire class. We def want isolation for backup jobs. 
- How will we handle fault tolerancy?  
    - Will we be able to replicate partitions across nodes? You touch upon it a bit, but I’m not sure if we cover it completely 
- How will we handle monitoring? 
    - We’ve talked about this shortly before, but being able to monitor and proactively alert will definitely be a topic that comes up with both Notion and OpenAI 
- Do we need to think about throttling? 
    - Not sure if this needs to be done on a Weaviate level, but I can imagine one user bashing their index can have a negative impact on the rest of the tenants. This might also not be for the first few iterations, but not sure if it’s a topic that needs to be discussed. Eg fair use policies etc..

**Comment by trengrj** [comment]
I like this approach. Some feedback:

- Users will probably discover they need partitioning once data has already been loaded into the cluster. So it would be nice to have the[ Reindex API ](https://github.com/weaviate/weaviate/issues/2359) available to them when this feature is available.
- As an additional ask it would be nice if we had a partitioning "property". Doing a where `Equal` filter on this property would have the same effect as specifying the `tenant`. Benefit of this approach is that users could migrate much easier as their GraphQL search queries would not need changing.

**Comment by etiennedi** [comment]
Thanks, both. These are all excellent points. Let me reply 1-by-1:

> How will we handle backups in these kinds of cases?
> The current backup functionality will not work, as it would backup the entire class. We def want isolation for backup jobs.

Agreed. The most straight-forward option here might be to add the field `partitions: []` to the backup API. It could behave similarly to the current class selector where you could backup (and restore) all partitions or only specific ones. Implementation-wise we might have to control this. With 1M tenants we probably don't want to backup all 1M shards in parallel (not enough network bandwidth). But we also don't want to do it sequentially (because it takes too long). So there is probably a good middle-ground of splitting this into smaller batches. (I'll create a dedicated Discovery issue for this)

> How will we handle fault tolerance?
> Will we be able to replicate partitions across nodes? You touch upon it a bit, but I’m not sure if we cover it completely

In my eyes, the same as before. If the class has a replication factor larger than `1`, every shard should be replicated according to the factor. Replicated shards cannot be scheduled on the same node as that would destroy fault-tolerance, but that is already the case with the current implementation. From a replication perspective we should already support everything, since the only change is that we have more shards, but @redouan-rhazouani can correct me if I'm wrong here.

In the first iteration a single replication factor per class should be sufficient. We could think about overriding the replication factor for specific partitions in the future. The motivation here would be a case where one tenant receives a lot of read-load and we want to replicate more to increase throughput. For now (MVP), I would consider this an edge case. 

> How will we handle monitoring?
> We’ve talked about this shortly before, but being able to monitor and proactively alert will definitely be a topic that comes up with larger customers

I've created a separate issue with the challenges and opportunities for monitoring in multi-tenancy environment: #2963. The tl;dr is that we need to drop the "per shard" cardinality, but at the same time we can use histograms to get good insights. For example, rather than measuring the latency for every tenant, we could use histograms to see "how many tenants have requests in the latency range of 50-100ms?"

> Do we need to think about throttling?
> Not sure if this needs to be done on a Weaviate level, but I can imagine one user bashing their index can have a negative impact on the rest of the tenants. This might also not be for the first few iterations, but not sure if it’s a topic that needs to be discussed. Eg fair use policies etc..

Good point. At the moment throttling would be happening at the class (or even Weaviate?)-level. This way one tenant could occupy the whole system. We probably need two kinds of limits:
* overall - to prevent the node from a DoS-like situation - and 
* per-tenant to make sure that a single tenant cannot "steal" resources from other tenants
There could also be hard limits, such as maximum number of goroutines per tenant. But I would also see this more as something to figure out down the line. Once we have a working multi-tenancy setup to run experiments, we can optimize for this.

> Users will probably discover they need partitioning once data has already been loaded into the cluster. So it would be nice to have thehttps://github.com/weaviate/weaviate/issues/2359available to them when this feature is available.

Agreed that this would be ideal, but I don't think it's realistic. To get the first iteration of partitioning ready in v1.20 we need to focus the entire team on this, there simply wouldn't be any capacity to tackle a Reindex API at the same time.

 > As an additional ask it would be nice if we had a partitioning "property". Doing a where Equal filter on this property would have the same effect as specifying the tenant. Benefit of this approach is that users could migrate much easier as their GraphQL search queries would not need changing.

I like this idea. Weaviate could rewrite the query under the hood to replace a where filter with a partition value, etc. A possible downside would be that this could complicate other APIs. Right now they're quite simple, with just (using HTTP query vars as an example) `?partion=foo`. If the portion becomes a key-value pair we might need something like `?partition=my_particionkey=foo`.

**Comment by trengrj** [comment]
> I like this idea. Weaviate could rewrite the query under the hood to replace a where filter with a partition value, etc. A possible downside would be that this could complicate other APIs. Right now they're quite simple, with just (using HTTP query vars as an example) ?partion=foo. If the portion becomes a key-value pair we might need something like ?partition=my_particionkey=foo.

We could set the property name in the Class schema via an option:

```
{
  /* ... other fields ... */
  "partitioning": {
    "enabled": true,
    "property": "tenant_id"
  }
}
```

So in your second example it would still be just `partition=foo`. I think this feature though could be added afterwards as is more a UX thing.

**Comment by etiennedi** [comment]
Thanks for the feedback, @modi95!

> Debugging Performance Issues
> It would be really useful to know if specific partitions are performing poorly. This may be because a there are multiple hot partitions on the same shard. Or perhaps a lot of slow queries involve some small subset of partitions.

Great point. For some of the current challenges around Metrics, see also #2963. An old-school "slow log" feature could also be super handy here. The cost of constantly observing every single tenants interaction would be extremely high, but if we define certain limits and only start observing/logging when a threshold is crossed this could be manageable. For example the metrics could indicate the number of tenants outside the desired latency window and then a slow log could log the details of which those tenants are, what node they're scheduled on, etc. 

> Volume Distribution across Partitions
> Some of our partitions will contain >10M vectors. And there will be a long tail of partitions with <10k vectors. This long tail can potentially be millions of partitions. Do you think this proposal can handle this long tail or perhaps this is a followup optimization?

Yes to both. It is specifically designed to be able to cope with many small partitions. At the same time, we'll keep optimizing it in the future. The main challenges to comfortably support such a long time are:

1. Maximize the number of partitions per node. We anticipate this to end up somewhere around 100k per node. With 10k vectors per partition, that would already be 1bn vectors per node. Chances are that we'd run out of resources on a single node not because it's too many tenants but because it's too many vectors. That would mean we've achieved our "many partitions per node" goal. The mentioned (future) improvements of offloading partitions that are currently inactive will also help tremendously here, as an inactive partition doesn't cost anything besides disk space. Some GH issues related to "maximizing the number of partitions per node" include #2947, #2983 ,#2951 
2. Make sure the number of partitions scale linearly in a multi-node cluster. If that's the case then whatever the per-node number of (1) is we can simply increase it by adding more nodes to the cluster. Issues related to that include #2948. I expect that once we have the MVP out and know where the scaling bottlenecks are, this is something that we can also really focus on in the next iteration.

> Remove inactive partitions from expensive “hot” nodes and offload them to cloud storage
> This one seems super useful!

Thanks, agreed. The idea of only paying for resources for currently active users, rather than paying for all users should lead to some significant cost savings down the line.

**Comment by redouan-rhazouani** [comment]
That’s a really nice approach to multi-tenancy. Some feedback:

**Dynamic Rebalancing:**

In a multitenant environment, we typically don't have a uniform distribution of data. If we simply map tenant IDs to partitions, we might end up creating hot/large partitions, which can impact the performance and cost of the application.
To solve such a problem, many DBMSs (DynamoDB, MongoDB, Cassandra) impose some limits on the partition size or the number of items in a partition, or both. They split the partition into two halves once the limit has been reached. Therefore, it makes more sense to design the system to allow for more than one partition per tenant.

**The Challenge**

I consider the schema update to be the biggest challenge in general. As described in the proposal, the current process for updating the schema doesn't scale well because it wasn't designed for such a large scale. Therefore, I think we need v1.20 to mainly solve this challenge. I would expect v1.20 to be focused around the following work packages:
1. Scalable schema update, which allows for:
- [ ] Concurrent update of independent classes.
- [ ] Concurrent update of different tenants (required for multi-tenancy).
- [ ] Update that doesn't require all nodes of the cluster to be available.
- [ ] Update that can tolerate node failure.
- [ ] Each node knows how to get the most recent schema of a specific class.

2. Efficient shard allocation:
- [ ] Minimise the cost of single shards in terms of memory and compute.

3. Support a minimal set of endpoints for the purpose of evaluation:
- [ ] Support for adding classes with multi-tenancy features enabled.
- [ ] Support for adding new tenants.
- [ ] Support for adding objects to new tenants.
- [ ] We may consider making these endpoints part of API-V2.

4. Evaluation:
- [ ] Evaluate the new implementation against the current one for multi-classes use case.
- [ ] Evaluate the new approach for multi-tenancy against the multi-classes approach.
- [ ] Determine how many shards can be supported in both approaches.
- [ ] Determine the optimal shard size.
- [ ] Determine the impact of the new implementation on performance and efficiency in general.

**Comment by etiennedi** [comment]
Thanks, @redouan-rhazouani, these are all excellent points. Let me respond per category:

> They split the partition into two halves once the limit has been reached. Therefore, it makes more sense to design the system to allow for more than one partition per tenant.

Agreed. That was the original motivation behind the virtual shards in a physical shard. The major blocker in ever actually implementing a shard-split was the insane cost of splitting a n HNSW vector index. A primitive approach would take roughly half of the original build time per half. So if a very large index was built for 10h, then the compute time to split it in half would be roughly 2x5h. That said, I'm sure there are better ways than such a primitive split, but they would themselves need to be researched. So my suggestion would be that we don't put too much focus on this right now, since in the known use cases, e.g. see @modi95's comment above, tenants tend to be in the range of 10k to 10M+ objects+vectors per shard. We have plenty of 100M+ single-shard setups running fine, so we know that this isn't a blocker which would give us some leeway to postpone any engineering efforts on splitting and rather focus on the long-tail of tenants for now.

> Scalable schema update, which allows for:
> * Concurrent update of independent classes.
> * Concurrent update of different tenants (required for multi-tenancy).
> * Update that doesn't require all nodes of the cluster to be available.
> * Update that can tolerate node failure.
> * Each node knows how to get the most recent schema of a specific class.

All great points, but I'm wondering if we need all of them in the first iteration already. Since most users would only have a single class (with millions of partitions), the benefit of concurrently updating independent classes wouldn't help too much here. 

I would suggest the following: If we can split out the "shard-assocation logic" and the new "partition state" from the schema (currently it's all in one), then we could build only this new logic (green field) to match the requirements you mentioned. This way, a user like @modi95 (haha, sorry, I'm using you as a stand-in for all multi-tenancy cases, but I do believe that yours is a very typical one) would have to create a class just once: It doesn't matter if that is slow and not fault-tolerant, because you only ever do it once. Then for every new  tenant that is onboarded this can be done concurrently and with fault-tolerance. 

Overall, I think we should prioritize an MVP use case over a perfect solution. I.e. it is important that we can eventually achieve fault-tolerance when adding new nodes, but I don't need that in the first release. The most important part for the first release is scale, so that users that a large number of tenants can get started with v1.20 – rather than having to wait for v1.21. In other words, it's better to support millions of tenants in v1.20, but only add fault tolerance in v1.21 than supporting only 100k tenants in v1.20, but fully fault-tolerant. 

(For people without context: When we talk about fault-tolerance only in the context of adding/removing tenants. Once a tenant exists, there is already fault-tolerance through the existing replication solutions).

> Efficient shard allocation:
> Minimise the cost of single shards in terms of memory and compute.

Agreed. The nice thing here is that this requires different skill sets and touching different parts of the code, so we can nicely parallelize this and the whole team can contribute to this feature.

> Support a minimal set of endpoints for the purpose of evaluation:

Similar to the MVP spirit above, I'm fine if we don't have all endpoints ready yet right away, as long as a typical user journey can be tackled in v1.20. Ideally, we should have a plan of how the API evolves though. If we cannot give stability guarantees for the API in v1.20 yet, then a user might have to wait for v1.21 until they can start using our new solution.

> Evaluation:
> * Evaluate the new implementation against the current one for multi-classes use case.
> * Evaluate the new approach for multi-tenancy against the multi-classes approach.
> * Determine how many shards can be supported in both approaches.
> * Determine the optimal shard size.
> * Determine the impact of the new implementation on performance and efficiency in general.

Great points. I can share a bit more about the limits of the multi-class solution. With just 20,000 classes, the time to rebuild the GraphQL schema was over 20s (!). That alone rules out multi-class for millions of tenants in my opinion. @modi95 can also share a few more insights about why using classes feels more like a workaround from a UX perspective and a partitioning (some competitors call it namespacing) solution that is easier to use for our users.

Besides this, I very much agree: Before we release v1.20 we should be very confident about how many tenants we can comfortably support and where the performance and efficiency pain points lie. We don't need to hyper-optimize everything in v1.20 yet, but we need to prove that this is the right solution and there are no architectural blockers from reaching the desired scale.

**Comment by dirkkul** [comment]
with regards to Search Operations (GraphQL/gRPC):

Instead of adding the tenant to the search operation, could we 
- Control with with user permissions which tenants can be accesed?
- Add an optional "Restrict to the following tenants" option to only search the given ones?

I can see usecases where users have access to multiple tenants, for example their personal-documents tenant and one or more team-document tenants that should be searched at the same time.

**Comment by etiennedi** [comment]
> I can see usecases where users have access to multiple tenants, for example their personal-documents tenant and one or more team-document tenants that should be searched at the same time.

One of the design principles above was that this should never be possible, i.e. strict isolation between tenants. I'm curious how users would model that. Maybe @modi95 can shed some light as well. I could imagine that a common approach would be to have the "organization" (or "team", "group", etc) be the tenant and then have individual user permissions regulated through specific fields and filters.

 @byronvoorbach might also have some insights into how other multi-tenancy cases treat this. I do remember that the user/group topic came up in a customer conversation at some point.

**Comment by ju-bezdek** [comment]
This would be indeed an awesome feature... now the need to use classes for separation breaks the concept that classes are the thing that describes the model. It is tedious when handling references, you can't have your queries predefined as you need to augment them by tenant id.

There is one good consequence of this approach however and that is the ease of offboarding tenant / deleting the data. So command like drop/delete partition would be nice to have to preserve this use case.

As for @trengrj  proposal on auto-partition definition by the property, that would be neat too, however, I'd prefer to have it as an optional feature. 

One more thing to consider, in enterprise databases like Oracle it's a common feature to partition by time (kind of like auto-archive)... that means that you can move older data to a cheaper partition. This might be useful when you do 90% of queries over the last 30d data ... but once in a while, you need to also older data... how Oracle handles it does search only in the current partition if the query contains a condition that clearly restricts the search for the partition, otherwise it does search everywhere. This is however advanced scenario... I personally don't have such need and it is probably less likely usecase for vector DB.

**Comment by vikasmehta-ib** [comment]
Overall, the spec looks really good and something we really need (have been looking into incrementally building this outside of Weaviate, really excited that we can avoid this). 

Can we break "step 1" in the timeline to first support this at single node level and later add multi-node support? Asking for solving single node problem first assuming it is a strict subset (building block) of the overall work and can be delivered faster. Also if consumer/customer are able to bring their own sharding / multi-node solution, it can help unblock until complete out-of-box horizontal scaling solution is ready.

While trying to store each tenant independently (class), first problem we have been struggling with is running a single Weaviate instance with large (10s of thousands) of small classes. We are trying to ensure that we can have a good control over physical resource usage (esp. memory) for a single Weaviate instance/node by keeping the working set of classes controlled (delete (ideally unload) in-active classes when we are reaching close to danger zone).

Trying to keep the in-memory working set of classes small (active users) by offloading classes using backup+delete class, some single node problems (we ran into):

1. Ability to create/delete classes concurrently - found a [global lock](https://github.com/weaviate/weaviate/blob/48456a16abba48abbc12d4ee72c4f2b0d5732952/usecases/schema/add.go#L119-L122) that serializes these API calls.
2.  Concurrent backup/restore that was mentioned in the spec.
3. Fix issues like delete class memory leaks (we found couple of leaks in delete_class - [compressedSharedLockCache](https://github.com/weaviate/weaviate/issues/2981) and prometheus stats module memory keeps growing as projects are created/deleted keeping number of active projects constant (will create a separate github issue for this)).

**Comment by nvbkdw** [comment]
Thanks for adding multi-tenancy to the roadmap, I want to add another two issues preventing us from using current Weaviate in multi-tenant environments: 
* addClass has a single lock, 
* and rebuildGraphQL schema latency grow linearly with the number of class in the system

I add more detail here: https://github.com/weaviate/weaviate/issues/2999

**Comment by eostis** [comment]
What will happen during an upgrade? Will tenants be impacted by a shutdown?

**Comment by meitalbensinai** [comment]
Hi, do you know to tell when this feature will be ready? 
thanks!

**Comment by nstiurca** [comment]
My feature request for post 1.20 enhancements: #3242

**Comment by dirkkul** [comment]
I think this is all implemented
